src/layout.rs

```diff
@@ -366,126 +382,139 @@ impl Layout {
+                solver
+                    .add_constraints(constraint.to_cassowary(element.width, area.width).iter())?;
+            }
         }
         Direction::Vertical => {
+            // ensure the elements touch the left and right edges of the area
+            for element in &elements {
+                solver.add_constraints(&[
+                    element.left() | EQ(REQUIRED) | area.left(),
+                    element.right() | EQ(REQUIRED) | area.right(),
+                ])?;
+            }
+            // ensure there are no gaps between elements
             for pair in elements.windows(2) {
-                ccs.push((pair[0].y + pair[0].height) | EQ(REQUIRED) | pair[1].y);
+                solver.add_constraint(pair[0].bottom() | EQ(REQUIRED) | pair[1].top())?;
             }
-            for (i, size) in layout.constraints.iter().enumerate() {
-                ccs.push(elements[i].x | EQ(REQUIRED) | f64::from(dest_area.x));
-                ccs.push(elements[i].width | EQ(REQUIRED) | f64::from(dest_area.width));
-                ccs.push(match *size {
-                    Constraint::Length(v) => elements[i].height | EQ(MEDIUM) | f64::from(v),
-                    Constraint::Percentage(v) => {
-                        elements[i].height | EQ(MEDIUM) | (f64::from(v * dest_area.height) / 100.0)
-                    }
-                    Constraint::Ratio(n, d) => {
-                        elements[i].height
-                            | EQ(MEDIUM)
-                            | (f64::from(dest_area.height) * f64::from(n) / f64::from(d))
-                    }
-                    Constraint::Min(v) => elements[i].height | GE(MEDIUM) | f64::from(v),
-                    Constraint::Max(v) => elements[i].height | LE(MEDIUM) | f64::from(v),
-                });
-
-                match *size {
-                    Constraint::Min(v) | Constraint::Max(v) => {
-                        ccs.push(elements[i].height | EQ(WEAK) | f64::from(v));
-                    }
-                    _ => {}
+            // ensure the first element touches the top edge of the area
+            if let Some(first) = elements.first() {
+                solver.add_constraint(first.top() | EQ(REQUIRED) | area.top())?;
+            }
+            // ensure the last element touches the bottom edge of the area. This constraint may not
+            // be satisfied if the constraints are impossible to satisfy.
+            if layout.expand_to_fill {
+                if let Some(last) = elements.last() {
+                    solver.add_constraint(last.bottom() | EQ(STRONG) | area.bottom())?;
                 }
             }
+            // apply the constraints
+            for (constraint, element) in layout.constraints.iter().zip(elements.iter()) {
+                solver
+                    .add_constraints(constraint.to_cassowary(element.height, area.height).iter())?;
```

Comment:
I like the refactor. 

I do have a mild preference for the `to_cassowary` method to be inlined here.

Reply:
I've looked at this both ways, and I think subjectively readability suffers when this is inlined. I wasn't sure if there was a better name or place to put the conversion however.

The main benefit on readability by putting this in a method is it's easy to see that the horizontal and vertical algorithms are essentially just switching top/bottom/left/right for left/right/top/bottom. This halves the cognitive load of the function.

I thought about some ways to condense the horizontal and vertical into a single function (e.g. by aliasing top/bottom/left/right as start/end/cross_start/cross_end), but can't find a way that keeps it simple, neat and obvious.

Reply:
Actually, I found a way to do both that works nicely, (by making `Element` direction aware and adding `start()`, `end`, etc. to pick the right variable.